Define Local Alleles in VCF to allow for sparser format (#434)
samtools/hts-specs: VCFv4.3.tex

      GQ		& 1		& Integer	& Conditional genotype quality \\
      GT		& 1		& String	& Genotype \\
      HQ		& 2		& Integer	& Haplotype quality \\

>       LAA		& . 		& Integer	& Local Alternate Alleles the 1-based index into the alternate alleles indicating which are relevant for the current sample.\\

Reviewer: ```suggestion
      LAA		& . 		& Integer	& Local Alternate Alleles the strictly-increasing, 1-based indices into the alternate alleles indicating which are relevant for the current sample.\\
```

Reviewer 2: Or maybe:
```suggestion
      LAA		& . 		& Integer	& 1-based index indicating which alternate alleles are relevant (local) for the current sample\\
```

Reviewer: SGTM but let's say the indices have to be strictly increasing; obviously that's a pedantic constraint but without it, the ordering of LPL could be unclear as well.

Reviewer 2: Suggestion updated.

Reviewer: Dwelling on this further -- @yfarjoun @lbergelson -- I'm worried just slightly about confusion over the LPL ordering. Lets say the joint VCF has REF=A ALT=G,C,T and the sample GVCF has ALT=T,G. so LPL=1,3 if we require the indices to be in order, but this order is flipped with respect to the GVCF.

The merging algo then has to be really careful to [permute the GVCF PL values](https://github.com/dnanexus-rnd/GLnexus/blob/237e9bcbcd685445ab0aa65ecbc6c29cefe1b26e/src/diploid.cc#L16-L43) in just the right way to match that. In the wild, I'm nervous this is going to lead to difficult-to-detect bugs with subtle swapping of probability values. Did you have any experience with this in developing the idea? Should we permit the merger to write an out-of-order LAA (this might just shift the potential confusion to readers)? I don't have a better idea right now, it just seems to me like a significant pitfall at least warranting explication where LPL is discussed.

Reviewer 3: I agree that this is a concern.  The only actual implementation so far of a format with local alleles decided to store the allele string itself instead of the index because they didn't want to deal with the reordering problem.  So unordered [T, G] instead of [1, 3]. Move the difficulties from the merge step to the reading step in that now if you want to look up non-local values associated with an allele you have to do a search in the INFO allele values to find the corresponding index.  I

My thinking was that it was better to require the indexes as numbers and do the reordering, but I think there's a strong argument the other way.

Reviewer: Thanks, I'll spawn a background thread on this. BTW, if anyone knows a better formula than the following to go from PL index `gt` to the indices `(i,j)` of the constituent alleles, I would like to know. This is seriously the best I could come up with, holy sh**.

```
    unsigned j = (unsigned)((sqrt(8*gt+1)-1.0)/2.0);
    unsigned i = gt - j*(j+1)/2;
```

Reviewer 4: I like @cyenyxe's suggestion, it solves a possible confusion. It also allows to determine the LPL ordering per allele-set, samples with the same set of local alleles are guaranteed to use the same order.

@mlin Note that the proposed formula is for the special case of diploid genotypes. In my experience, in practice it is sufficient to pre-compute the indexes using nested/recursive loops as suggested in the description of GL, section 1.6.2.

Author: why do the indexes need to be strictly increasing? 

One of the main points is to lessen the burden on the creation of these pVCFs...so that when the alleles change the only change that needs to happen in the format field is the LAA. if we require strictly increasing order, that goes out the window.

Reviewer 4: If LAA are strictly increasing, LPL's are guaranteed to have the same order as 1) the original PLs and 2) samples with the same set of alleles.

Editing of VCFs is much less frequent operation than read-only processing, therefore it is more important to optimize the speed and convenience of the latter.

>       LPL		& . 		& Integer	& Phred-scaled genotype likelihoods rounded to the closest integer for genotypes that involve the Reference and the LAA alleles only.\footnotemark[1]\\ 

Reviewer 3: Shouldn't LPL be sorted after LAD?

Reviewer 4: Both LPL and LAD should be sorted after REF+LAA, the same way as REF+ALT define the order of PL and AD.

>       LAD		& . 		& Integer	& Local Allele Depth for the reference and each of the local alternate alleles (see: LAA).\\

Reviewer 2: ```suggestion
      LAD		& . 		& Integer	& Read depth for the reference and each of the local alternate alleles listed in LAA\\
```

>       LPL		& . 		& Integer	& Phred-scaled genotype likelihoods rounded to the closest integer for genotypes that involve the Reference and the LAA alleles only (see LAA).\\

Reviewer 2: ```suggestion
      LPL		& . 		& Integer	& Phred-scaled genotype likelihoods rounded to the closest integer for genotypes that involve the reference and the local alternate alleles listed in LAA\\
```

Reviewer: ```suggestion
      LPL		& . 		& Integer	& Phred-scaled genotype likelihoods rounded to the closest integer for genotypes that involve the reference and the local alternate alleles listed in LAA, in the genotype order implied by LAA such that LPL would be a subsequence of PL were the latter present.\\
```

@lbergelson @pd3 following the previous discussion, I feel it's worth leaving no possible question about the LPL order -- here's my attempt, better suggestions welcome.

For GVCF merging, we're already responsible for carefully permuting the GVCF PL values into the target joint vector, the new wrinkle will be that each sample can have a different target. I think this is okay, it was just tempting to want to copy in the GVCF PL verbatim.

Reviewer 4: I think it's better to speak about ordering only and not require a strict subsequence: it should be possible to rescale the LPL likelihoods.

      GQ		& 1		& Integer	& Conditional genotype quality \\
      GT		& 1		& String	& Genotype \\
      HQ		& 2		& Integer	& Haplotype quality \\

>       LAA		& . 		& Integer	& Strictly increasing, 1-based indies into ALT, indicating which alternate alleles are relevant (local) for the current sample \\

Reviewer 3: typo: *indies*

  All phased genotypes that do not contain a PS subfield are assumed to belong to the same phased set.
  If the genotype in the GT field is unphased, the corresponding PS field is ignored.
  The recommended convention is to use the position of the first variant in the set as the PS identifier (although this is not required).

>   \item LAA (and LAD and LPL (*):

Reviewer 2: I would follow the existing of one item per field, otherwise it may be confusing to readers searching for LAD and LPL.

Author: should I leave the paragraph as is, and add entries for LAD and LPL pointing to LAA?

Reviewer 5: The AD, ADF, ADR paragraph also groups related items into a single paragraph. Fortuitously in both cases the three entries are adjacent alphabetically.

(Though the whole LAA/LAD/LPL paragraph wants to move up to between HQ and MQ, and this label (`LAA (and LAD and LPL (*)`) wants less ‘and’ and better balanced parentheses!)

  For callsets with a large number of samples, it is often the case that the majority of sites are not called and sites end up involving many alleles for which all the samples need to provide PL and AD. 

>   This can cause the file-sizes to grow super-linearly with the number of samples. 

Reviewer: Different take on the previous two sentences; feel free to use/recombine

> In callsets with many samples, sites may grow to include numerous alternate alleles at the same POS. Usually, few of these alleles are actually observed in any one sample, but each sample must supply fields like PL and AD for all of the alleles -- a very inefficient representation as PL's size is quadratic in the allele count.

  All phased genotypes that do not contain a PS subfield are assumed to belong to the same phased set.
  If the genotype in the GT field is unphased, the corresponding PS field is ignored.
  The recommended convention is to use the position of the first variant in the set as the PS identifier (although this is not required).

>   \item LAA is a sorted list of n distinct integers, where 1 <= n <= $|ALT|$, giving the (1-based) indices within ALT of the alleles that are observed in the sample. 

Reviewer 6: **1 <= n <= $|ALT|$** converts to **1 ¡= n ¡= |ALT |** in the PDF.  Is this what was intended? (Not a notation I have seen before if so).

Reviewer 5: PR #745 corrects the formatting of the mathematics.

  In this case LGT=0/1 means that the sample is G/C. 
  GQ is still the genotype quality, even when the genotype is given against the local alleles. 
  Note that reordering might be required and care need to be taken to reorder LAD and LPL appropriately.
  LAA is required in order to interpret LAD, LPL, and LGT.
  
  In the following example, the records with the same POS encode the same information (some columns removed for clarity):
 
  \begin{tabular}[l]{llllll}
POS &REF&  ALT&FORMAT&sample\\
1&G&A,C,T,\textless*\textgreater& LAA:LGT:LAD:LPL&  2,4:1/1:20,30,10:90,80,0,100,110,120\\
1&G&A,C,T,\textless*\textgreater& GT:AD:PL&  2/2:20,.,30,.,10:90,.,.,80,.,0,.,.,.,.,100,.,110,.,120\\
2&A&C,G,T,\textless*\textgreater& LAA:LGT:LAD:LPL& 3:0/1:15,25:40,0,80\\
2&A&C,G,T,\textless*\textgreater& GT:AD:PL&0/3:15,.,.,25,.:40,.,.,.,.,.,0,.,.,80,.,.,.,.\\
3&C&G,T,\textless*\textgreater& LAA:LGT:LAD:LPL& 4:0/0:30,1:0,30,80\\
3&C&G,T,\textless*\textgreater& GT:AD:PL& 0/0:30,.,.1:0,.,.,.,.,.,.,.,.,.,30,.,.,.,80\\

> 4&G&A,T,\textless*\textgreater& LAA:LGT:LAD:LPL& :0/0:30:0\\

Reviewer 3: Typo here:  we either need to remove the LAA field and indicate that it can be left out if there are no non-ref local alleles, or we need to add a leading `.`

Reviewer 5: I don't think this is a typo: the intention is that LAA is a 0-length list of integers (indicating that this record is homozygous ref for this sample and that none of the ALT alleles are in play), not an unknown MISSING value.

This may be a problem, as it's not clear that 0-length lists are valid syntax in VCF or BCF. Omitting LAA is probably not the right workaround, as presence of LAA is the simple indicator that local alleles are being used in a VCF record. Setting it to `.` (unknown) is somewhat misleading, as it **is** known…

Reviewer 5: This example contradicts the current text, which says “LAA is a sorted list of _n_ distinct integers, where _n_ ≥ 1 […]”. So the text will need adjustment if this example record is deemed valid.

Reviewer 7: `bcftools` converts the zero-length list into a missing value.  This comes from bcftools treating everything as BCF internally, and [BCF says zero-length lists are missing](https://samtools.github.io/hts-specs/VCFv4.4.pdf#subsubsection.6.3.3).

Possible fixes would be to say missing `LAA` means REF; to explicitly put 0 in `LAA` for REF-only; or use `GT` in these rows, but allow use of `LPL`, `LAD` etc. in the specific case that `GT` is REF-only.

Reviewer 4: It makes sense to use the missing value `.` and state it explicitly that it means REF. I can't think of a case where this could cause problems.

Reviewer 8: I agree with Petr here.  While the BCF specification states that a zero length array is defined to be a MISSING array, hence `.`, we could add an explicit statement to LAA saying that missing LAAs means that it matches REF.  What we gain is a solution to this problem.  What we lose is the ability to store an unknown LAA list.  I'm unsure what the latter practically means.  Why would you need to explicitly state it's missing instead of just omitting it.  If it's missing then we're not using LAA (and if the PL, AD etc are missing we're not defining it that way either).

PS. Technically even if it's `.` this still contradicts "LAA is a sorted list of n distinct integers" as `.` isn't an integer.  However that's nit-picking and maybe somewhere else VCF already defines `.` as being an integer of unknown value?  I don't know what we do elsewhere when we have this technical disparty.  Is it just assumed that "." is a polymophic type that matches whatever the item wants it to be?

Reviewer 5: What the BCF spec says is that a type descriptor byte with ‘size’=0 denotes MISSING. It does not say what a type descriptor byte with ‘size’=15 followed by a typed Integer with value 0 denotes, but that would be a natural representation of a 0-length list. (The spec does say that ‘size’=15 implies the list length is ≥15, but the typed Integer is not biased so no meanings are described for typed Integer values <15.) Thus BCF does not provide a binding answer on whether 0-length lists are valid or representable in VCF/BCF.

But I'm not going to advocate blessing 0-length lists in VCF or BCF, especially not just for this, even though SAM found them useful for regularity (cf #135).

IMHO the clearest way to represent this corner case would be

> As a special case, the homozygous REF case in which no ALT alleles are observed in a particular sample is represented as LAA=[0].

(And we'd need to rework _n_ in the length maths: e.g. by dropping _n_ from the description of LAA, and adding “where _n_ is the number of **ALT** alleles listed in LAA” to the others.)

Reviewer 4: To expand on my previous comment: the logic of LAA is to define a list of alleles that can be referred to in subsequent localized fields. If only the reference allele is relevant, then it makes sense to provide the missing value `.`, in this sense it behaves in exactly the same way as what we write in REF and ALT columns. Note also the naming of the tag stands for "Local Alternate Alleles". By using `.` we are not introducing a new concept.

Reviewer 9: Parachuting in here a bit, so my apologies if I'm missing context, but defining LAA as an R-length array whose first element is always the number zero both avoids these nasal demons and avoids a branch [1] when connecting a local index into a global index.

Is it too late to consider an R-length LAA instead of an A-length one?

[1] Text processing cost probably dominates, but if you convert a VCF into some binary format without adjusting LAA, maybe this matters a bit.
